### ISSUE TITLE
Wrong output shape with 0 padding for transposed convolution.

### ISSUE THREAD

**Issue by windsorwho** [google/jax#17618]
### Description

```
import jax
import jax.numpy as jnp
latent = jnp.ones((1, 128,128,128)) # NHWC
kernel = jnp.ones((4,4,3,128)) # K, K, out_chn, in_chn
image = jax.lax.conv_transpose(lhs=latent, rhs=kernel, strides=(4,4), 
                               padding='VALID', rhs_dilation=None, 
                               dimension_numbers=("NHWC", "HWIO", "NHWC"),
                               transpose_kernel=True)
image.shape
```
image shape is (1, 512, 512, 3)
But if I change the padding to [(0,0),(0,0)]
```
import jax
import jax.numpy as jnp
latent = jnp.ones((1, 128,128,128)) # NHWC
kernel = jnp.ones((4,4,3,128)) # K, K, out_chn, in_chn
image = jax.lax.conv_transpose(lhs=latent, rhs=kernel, strides=(4,4), 
                               padding=[(0,0),(0,0)], rhs_dilation=None, 
                               dimension_numbers=("NHWC", "HWIO", "NHWC"),
                               transpose_kernel=True)
image.shape
```
image shape is (1, 506, 506, 3).
My understanding is that the output shape should be the same as no padding,
Am I wrong?

### What jax/jaxlib version are you using?

jax 0.4.13

### Which accelerator(s) are you using?

CPU/TPU

**Comment by jewillco** [comment]
The padding applied internally for `VALID` padding is not zero; it seems to be `(3, 3)` in both spatial dimensions from reading through the source code. The relevant code is [here](https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/convolution.html#conv_transpose); scroll up for the helper functions. Transposed convolutions with strides are implemented internally as forward convolutions with the input image dilated by the given strides (and other aspects changed), and so extra padding is needed to make the result shape match the forward pass's input shape. In your case, the kernel needs to be applied four times to each input pixel in each dimension; a 1-D version of the needed `VALID` padding is:

<table width="100%">
<tbody>
<tr><td>⚪</td><td>⚪</td><td>⚪</td><td>🔴</td><td>⚪</td><td>⚪</td><td>⚪</td><td>🔴</td><td>⚪</td><td>⚪</td><td>⚪</td><td>🔴</td><td>⚪</td><td>⚪</td><td>⚪</td><td>🔴</td><td>⚪</td><td>⚪</td><td>⚪</td></tr>
<tr><td>⬛</td><td>⬛</td><td>⬛</td><td>⬛</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr>
<tr><td></td><td>🟦</td><td>🟦</td><td>🟦</td><td>🟦</td><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/></tr>
<tr><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td/><td>🟨</td><td>🟨</td><td>🟨</td><td>🟨</td></tr>
</tbody>
</table>

where the input pixels from your tensor are the 🔴's, the ⚪'s are inserted zeros from the dilation, and the ⬛⬛⬛⬛, 🟦🟦🟦🟦, and 🟨🟨🟨🟨 boxes show the extents of the kernel in a few of its possible positions. The kernel is scanned one position at a time along the dilated and padded image. This 1-D toy example produces 16 output pixels, which is what would be expected for a stride of 4 and a kernel size of 4.

**Comment by windsorwho** [comment]
Thanks for the clarification. I thought conv_transpose and conv are revisable given the same params, but it seems not, which is fine.
